src/PrestaShopBundle/Controller/ArgumentResolver/SearchParametersResolver.php

```diff
@@ -109,45 +110,101 @@ public function supports(Request $request, ArgumentMetadata $argument)
+        $this->dispatcher->dispatch(FilterSearchCriteriaEvent::NAME, $filterSearchParametersEvent);
+
+        yield $filterSearchParametersEvent->getSearchCriteria();
+    }
+
+    /**
+     * @param Filters $filters
+     * @param string $controller
+     * @param string $action
+     *
+     * @return bool Indicates if filters have been overridden
+     */
+    private function overrideWithSavedFilters(Filters $filters, $controller, $action)
+    {
+        if (!empty($filters->getUuid())) {
+            /** @var Filters $savedFilters */
+            $savedFilters = $this->searchParameters->getFiltersFromRepositoryByUuid(
+                $this->employee->getId(),
+                $this->shopId,
+                $filters->getUuid(),
+                get_class($filters)
+            );
+        } else {
             /** @var Filters $savedFilters */
             $savedFilters = $this->searchParameters->getFiltersFromRepository(
                 $this->employee->getId(),
                 $this->shopId,
                 $controller,
                 $action,
-                $filtersClass
+                get_class($filters)
             );
-
-            if ($savedFilters) {
-                $filters->add($savedFilters->all());
-            }
         }
 
-        //Then override with query filters if present
-        $query = $request->query;
-        $queryHasFilters = false;
-        foreach (SearchParametersInterface::FILTER_TYPES as $filterType) {
-            if ($query->has($filterType)) {
-                $queryHasFilters = true;
-                break;
-            }
+        if ($savedFilters) {
+            $filters->add($savedFilters->all());
+
+            return true;
         }
-        if ($queryHasFilters) {
-            /** @var Filters $queryFilters */
-            $queryFilters = $this->searchParameters->getFiltersFromRequest($request, $filtersClass);
+
+        return false;
+    }
+
+    /**
+     * @param Request $request
+     * @param Filters $filters
+     *
+     * @return bool Indicates if filters have been overridden
+     */
+    private function overrideWithRequest(Request $request, Filters $filters)
+    {
+        /** @var Filters $queryFilters */
+        $queryFilters = $this->searchParameters->getFiltersFromRequest($request, get_class($filters));
+        if ($queryFilters->count()) {
             $filters->add($queryFilters->all());
 
-            //Update the saved filters (which have been modified by the query)
-            $filtersToSave = $filters->all();
-            unset($filtersToSave['offset']); //We don't save the page as it can be confusing for UX
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param Filters $filters
+     * @param string $controller
+     * @param string $action
+     * @throws \Doctrine\ORM\OptimisticLockException
+     */
+    private function persistFilters(Filters $filters, $controller, $action)
```

Comment:
I feel like `SearchParametersResolver` has too many responsibilities. As name suggests, it should only resolve parameters wihout any _write_ operations.

Reply:
I agree with, i proposed a new implementation with `FiltersBuilderInterface`, it is more solid as it is composed of multiple builders (one for default class instantiation, one for request, one for repository fetch, one for repository persistance, one for event)

The idea behind this is that the interface is more generic (`setConfig`, `buildParameters`) so that future modifications can use the same interfaces without BC, you simply need to implement a new builder and change the service definition.

For now it has the same workflow as `SearchParametersResolver` (filters are saved in database via the resolver) but it can be changed once we are sure of the new workflow.